dali/pipeline/operators/attributes.h

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef DALI_PIPELINE_OPERATORS_ATTRIBUTES_H_
+#define DALI_PIPELINE_OPERATORS_ATTRIBUTES_H_
+
+#include <vector>
+#include <utility>
+#include "dali/pipeline/operators/common.h"
+#include "dali/pipeline/workspace/sample_workspace.h"
+
+namespace dali {
+
+class CastPermuteAttr {
+ protected:
+  explicit inline CastPermuteAttr(const OpSpec &spec, bool defaultCastPermut = true) :
+    image_type_(spec.GetArgument<DALIImageType>("image_type")),
+    C_(IsColor(image_type_) ? 3 : 1) {
+    if (defaultCastPermut) {
+      output_type_ = DALI_NO_TYPE;
+      output_layout_ = DALI_SAME;
+    } else {
+      output_type_ = spec.GetArgument<DALIDataType>("output_dtype");
+      output_layout_ = spec.GetArgument<DALITensorLayout>("output_layout");
+    }
+  }
+
+  void SetupSharedSampleParams(const DeviceWorkspace *ws) {
+    if (output_type_ == DALI_NO_TYPE)
+      output_type_ = ws->Input<GPUBackend>(0).type().id();
+  }
+
+  void SetupSharedSampleParams(const SampleWorkspace *ws) {
+    if (output_type_ == DALI_NO_TYPE)
+      output_type_ = ws->Input<CPUBackend>(0).type().id();
+  }
+
+  // Output data type
+  DALIDataType output_type_;
+
+  // Output data layout
+  DALITensorLayout output_layout_;
+  const DALIImageType image_type_;
+  const int C_;
+};
+
+class CropAttr : protected CastPermuteAttr {
```

Comment:
I think that is not a good idea. Crop operator will have to inherit from CropAttr, which as of now is not ideal. With this change it would also include the arguments for CastPermute, which takes us far from where we want to be. I think that refactoring all these args classes could be a task in itself, but I wouldnt like to keep piling up complexity on it. My proposition is to keep by now these guys separated and have CropCastPermute for example ,inherit from both separately. Ideally it would be good to remove this inheritance